File: OBI/2021/M2/Level_2/fast_calcule.py
def num_decompose(x:int)->list:
    num_decom=[];
    while(x):
        num_decom.append(x%10);
        x//=10;
    num_decom.reverse()
    return num_decom
def get_digits_sum_equal(num:list,trigget:int,level:int,sum_:int,equal:int)->int:
    if(sum_==equal):
        return 1
    if(level>=len(num) or sum_>equal):
        return 0

    range_max=num[level] if trigget else 9
    num_level=num[level]
    instances=0
    #print('(',end='')
    for i in range(range_max+1):
        #print(' {}'.format(i),end='')
        new_trigget=1 if i==num_level and trigget else 0
        new_sum=sum_+i
        instances+=get_digits_sum_equal(num,new_trigget,level+1,new_sum,equal)
    #print(')',end='')
    return instances

def main():
    s=int(input())
    a=int(input())
    b=int(input())

    a_decompose=num_decompose(a-1)
    b_decompose=num_decompose(b)

    a_digits=get_digits_sum_equal(a_decompose,1,0,0,s)
    b_digits=get_digits_sum_equal(b_decompose,1,0,0,s)

    print(b_digits-a_digits)

if __name__=="__main__":
    main()

File: OBI/2021/M2/level_2/fast_calcule.cpp
#include<algorithm>
#include<iostream>
#include<vector>

std::vector<int>number_decompose(int x){
    std::vector<int>num;
    while(x){
        num.push_back(x%10);
        x/=10;
    }
    std::reverse(num.begin(),num.end());
    return num;
}
int calcule_sum_digits_equal(std::vector<int>num,int trigget,size_t level,int sum,int equal){
    if(sum==equal)
        return 1;
    if(level>=num.size()||sum>equal)
        return 0;

    int range_max=(trigget)?num[level]:9;
    int num_level=num[level];
    int instance=0;
    for(int i=0;i<=range_max;++i){
        int new_trigget=(trigget&&i==num_level)?1:0;
        int new_sum=sum+i;

        instance+=calcule_sum_digits_equal(num,new_trigget,level+1,new_sum,equal);
    }
    return instance;
}
//
int main(){
    int s,a,b;
    std::scanf("%i",&s);
    std::scanf("%i",&a);
    std::scanf("%i",&b);

    std::vector<int>a_decompose=number_decompose(a-1);
    std::vector<int>b_decompose=number_decompose(b);

    int a_digits=calcule_sum_digits_equal(a_decompose,1,0,0,s);
    int b_digits=calcule_sum_digits_equal(b_decompose,1,0,0,s);
    
    printf("%i\n",b_digits-a_digits);

    return 0;
}
